src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java

```diff
@@ -404,74 +423,107 @@ public void jobFinished () {
             // First start graph-building instance and wait for graph to successfully build.
             status.message = "Starting up graph building EC2 instance";
-            List<Instance> instances = startEC2Instances(1);
+            List<Instance> instances = startEC2Instances(1, false);
+            // Exit if an error was encountered.
+            if (status.error || instances.size() == 0) {
+                ServerController.terminateInstances(getIds(instances));
+                return;
+            }
             status.message = "Waiting for graph build to complete...";
-            MonitorServerStatusJob monitorInitialServerJob = new MonitorServerStatusJob(owner, deployment, instances.get(0), otpServer);
+            MonitorServerStatusJob monitorInitialServerJob = new MonitorServerStatusJob(owner, this, instances.get(0), false);
             monitorInitialServerJob.run();
+
             status.update("Graph build is complete!", 50);
+            // If only building graph, job is finished. Note: the graph building EC2 instance should automatically shut
+            // itself down if this flag is turned on (happens in user data). We do not want to proceed with the rest of
+            // the job which would shut down existing servers running for the deployment.
+            if (deployment.buildGraphOnly) {
+                status.update("Graph build is complete!", 100);
+                return;
+            }
+            Persistence.deployments.replace(deployment.id, deployment);
+            if (monitorInitialServerJob.status.error) {
+                // If an error occurred while monitoring the initial server, fail this job and instruct user to inspect
+                // build logs.
+                statusMessage = "Error encountered while building graph. Inspect build logs.";
+                LOG.error(statusMessage);
+                status.fail(statusMessage);
+                ServerController.terminateInstances(getIds(instances));
+                return;
+            }
             // Spin up remaining servers which will download the graph from S3.
             int remainingServerCount = otpServer.ec2Info.instanceCount <= 0 ? 0 : otpServer.ec2Info.instanceCount - 1;
+            List<MonitorServerStatusJob> remainingServerMonitorJobs = new ArrayList<>();
+            List<Instance> remainingInstances = new ArrayList<>();
             if (remainingServerCount > 0) {
                 // Spin up remaining EC2 instances.
                 status.message = String.format("Spinning up remaining %d instance(s).", remainingServerCount);
-                List<Instance> remainingInstances = startEC2Instances(remainingServerCount);
-                instances.addAll(remainingInstances);
+                remainingInstances.addAll(startEC2Instances(remainingServerCount, true));
+                if (remainingInstances.size() == 0 || status.error) {
+                    ServerController.terminateInstances(getIds(remainingInstances));
```

Comment:
It's only 1 line of code, but it may still make sense to create a method to terminate instances given some IDs.